ethereum/contracts/query/QueryResponse.sol

```diff
@@ -0,0 +1,220 @@
+// contracts/query/QueryResponse.sol
+// SPDX-License-Identifier: Apache 2
+
+pragma solidity ^0.8.0;
+
+import "../libraries/external/BytesLib.sol";
+import "../interfaces/IWormhole.sol";
+
+/// @dev QueryResponse is a library that implements the parsing and verification of Cross Chain Query (CCQ) responses.
+abstract contract QueryResponse {
+    using BytesLib for bytes;
+       
+    /// @dev ParsedQueryResponse is returned by parseAndVerifyQueryResponse().
+    struct ParsedQueryResponse {
+        uint8   version;
+        uint16  senderChainId;
+        bytes   requestId; // 65 byte sig for off-chain, 32 byte vaaHash for on-chain
+        uint32  nonce;
+        ParsedPerChainQueryResponse [] responses;
+    }
+
+    /// @dev ParsedPerChainQueryResponse describes a single per-chain response.
+    struct ParsedPerChainQueryResponse {
+        uint16 chainId;
+        uint8 queryType;
+        bytes request;
+        bytes response;
+    }
+
+    /// @dev EthCallQueryResponse describes an ETH call per-chain query.
+    struct EthCallQueryResponse {
+        bytes requestBlockId;
+        uint64 blockNum;
+        bytes32 blockHash;
+        uint64 blockTime;
+        EthCallData [] result;
+    }
+
+    /// @dev EthCallData describes a single ETH call query / response pair.
+    struct EthCallData {
+        address contractAddress;
+        bytes callData;
+        bytes result;
+    }    
+
+    bytes public constant responsePrefix = bytes("query_response_0000000000000000000|");
+
+    /// @dev getResponseHash computes the hash of the specified query response.
+    function getResponseHash(bytes memory response) public pure returns (bytes32) {
+        return keccak256(response);
+    }
+
+    /// @dev getResponseDigest computes the digest of the specified query response.
+    function getResponseDigest(bytes memory response) public pure returns (bytes32) {
+        return keccak256(abi.encodePacked(responsePrefix,getResponseHash(response)));
+    }
+    
+    /// @dev parseAndVerifyQueryResponse verifies the query response and returns the parsed response.
+    function parseAndVerifyQueryResponse(address wormhole, bytes memory response, IWormhole.Signature[] memory signatures) public view returns (ParsedQueryResponse memory r) {
+        verifyQueryResponseSignatures(wormhole, response, signatures);
+
+        uint index = 0;
+        
+        r.version = response.toUint8(index);
```

Comment:
We should consider using the unchecked methods from this [bytes parsing library](https://github.com/wormhole-foundation/wormhole/blob/main/ethereum/contracts/relayer/libraries/BytesParsing.sol) to reduce overall gas usage.